File: 16.3-sum-closest.py
#
# @lc app=leetcode id=16 lang=python3
#
# [16] 3Sum Closest
#

# @lc code=start
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n=len(nums)
        MIN=nums[0]+nums[1]+nums[2]
        for i in range(n-2):
            left=i+1
            right=n-1
            while( left<right):
                sum=nums[i]+nums[left]+nums[right]
                if (abs(sum-target)<abs(MIN-target)):
                    MIN=sum
                if sum<target:
                    left+=1
                elif sum>target:
                    right-=1
                else :
                    return MIN
        return MIN 
'''双指针法的python解决办法 还存在优化的空间 比如nums[i]和nums[i-1]相同时可以直接跳过
# @lc code=end

File: 16.3-sum-closest.cpp
/*
 * @lc app=leetcode id=16 lang=cpp
 *
 * [16] 3Sum Closest
 */

// @lc code=start
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        int n=nums.size();
        int MAX=nums[0]+nums[1]+nums[2];
        for (int i=0;i<n-2;i++){
            int left=i+1;
            int right=n-1;
            while(left<right){
            int sum=nums[i]+nums[left]+nums[right];
            if (abs(target-sum)<abs(MAX-target))
                MAX=sum;
              if(sum > target)
                    right--;
                else if(sum < target)
                    left++;
                else
                    return MAX;

            }
        }
        return MAX;
    }
};
/*常规的双指针解法
// @lc code=end

